keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1503 @@
+
+The declarative validation rules will also be included in the published OpenAPI:
+
+```json
+// Pod Spec
+"openAPIV3Schema": {
+  "type": "object",
+  "x-kubernetes-validations": [
+    {
+      "rule": "!self.hostNetwork || self.containers.all(c, c.containerPort.all(cp, cp.hostPort == cp.containerPort))"
+    }
+  ],
+}
+
+...
+
+// Container Port
+"openAPIV3Schema": {
+  "type": "object",
+  "properties": {
+    "hostPort": {
+      "minimum": 1,
+      "maximum": 65535
+      ...
+    }
+    "containerPort": {
+      "minimum": 1,
+      "maximum": 65535
+      ...
+    }
+  }
+}
+```
+
+It is important to note that the declarative validation will use the same set of
+validation options available today for CRDs. namely:
+
+- [JSON Schema value validations](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00) (e.g. `format`, `maxItems`)
+- [CEL Validation Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) (e.g. `self.replicas <= self.maxReplicas`)
+- Relevant [merge strategy declarations](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) (e.g. `listType=set`)
+
+## Motivation
+
+Kubernetes API validation rules are currently written by hand, which makes them
+difficult for users to access directly.
+
+Declarative validation will benefit Kubernetes maintainers:
+
+- It will make it easier to develop, maintain and review APIs.
+- It will make it easier programatically inspect and analyze the API, enabling
+  new tools and improved documentation.
+- It will enable improvements to the API machinery. For example, a feature like
+  ratcheting validation will become more tractable to implement because the
+  feature can be implemented once in the declarative validation subsystem
+  rather than piecemeal across the 15k lines of hand written validation code.
+
+Declarative validation will also benefit Kubernetes users:
+
+- It will give users direct access to the actual API validation rules, which
+  are currently only available to developers willing and able to find and read
+  the hand written validation rules.
+- It will enable clients to perform validation of native types earlier in
+  development worflows ("shift-left"), such as at with a Git pre-submit linter.
+- It will improve API composition. In particular CRDs that embed native types
+  (such as PodTemplate), which gain validation of the native type automatically.
+  This has the potential to simplify controller development and improve end
+  user experiences when using CRDs.
+- Validation is performed on versioned type, so error message and attributed field
+  path is var more likely to be relevant to the user.
+
+Please feel free to try out the
+[prototype](https://github.com/jpbetz/kubernetes/blob/cel-for-native/staging/src/k8s.io/sample-apiserver/pkg/apis/wardle/validation/README.md)
+to get hands on experience with this proposed enhancement.
+
+### Goals
```

Comment:
I think it should be a goal that validation errors for CEL, OpenAPI types and hard-coded functions are ALL returned as one single list. It's a nightmare for users if we do have first errors for one thing and then errors for the next layer.

Reply:
added a note that client experience does not change. the current way errors are handled client side should not change even a little bit. Maybe there would be some extra work during alpha to find our "error difference logs" but I think thats different